### ISSUE TITLE
Github Actions to auto publish npm package

### ISSUE THREAD

**Issue by munanadi** [coral-xyz/anchor#1019]
Shouldn't we write a Github action to take care of pushing the npm anchor thing automatically? Like say parse a commit tag and if it says some version run like `0.18.2` that should trigger and push it to npm.

It could solve one part of 
![image](https://user-images.githubusercontent.com/22261173/141696897-74403d40-2524-4c58-9eb6-7521815bc4ec.png)

lol :stuck_out_tongue: 

The action config file can look something like this. Not sure if this will work though. Never done this before. Need to double-check.
Thoughts on the whole actions thing though?

```yml
name: Release

on:
  workflow_dispatch: {}
  push:
    tags:
      - "v*.*.*"

jobs:
  programs:
    runs-on: ubuntu-latest
    name: Build and publish the SDK
    steps:
      - uses: actions/checkout@v2
      - name: Setup Node
        uses: actions/setup-node@v2
        with:
          node-version: 14.x
          registry-url: "https://registry.npmjs.org"
      - run: yarn install
      - run: yarn build
      - name: Publish
        run: yarn publish ./ts/dist
        env:
          NODE_AUTH_TOKEN: ${{ secrets.NPM_PUBLISH_TOKEN }}
```

**Comment by tomlinton** [comment]
Good idea. I was going to propose something similar after #1010. It might look something like this:

- Trigger a workflow with a version number as a parameter
- Run the `version-bump.sh` script with that parameter, e.g. `./version-bump.sh <version>`
- Build and publish NPM package, as you've outlined
- Run `yarn install` in `examples/tutorial` and `tests` to update the JS lock files with the newly published NPM package.
- Publish Rust crates
- Commit all those changes
- Create tag for <version>

It might also makes sense to split this into multiple actions to insert a manual human review into the flow somewhere.

**Comment by fanatid** [comment]
It would be cool to have auto-publish packages in CI but I'd like that somebody checked security first.
- Is it possible to have a publish token for the npm package if 2FA is enabled?
- Is it possible to have a publish token for Rust crates?
- How it's secure to store tokens on GitHub?

**Comment by tomlinton** [comment]
> * Is it possible to have a publish token for the npm package if 2FA is enabled?

Yes, you can prompt for 2FA, e.g. https://github.com/mmarchini-oss/npm-otp-publish. It might be worth rolling our own or at least going through the source of that.

> * Is it possible to have a publish token for Rust crates?

Yes

> * How it's secure to store tokens on GitHub?

Details of how secrets are stored is here, https://docs.github.com/en/actions/security-guides/encrypted-secrets. They are meant to be encrypted throughout the process until they are used.

**Comment by fanatid** [comment]
`npm-otp-publish` is awesome, hope that crates.io also introduces 2FA someday.

**Comment by Henry-E** [comment]
This might be cool but I agree there are security issues, so for now we might stick to a manual checklist.
